lib/view_component/test_helpers.rb

```diff
@@ -34,20 +36,22 @@ def render_inline(component, **args, &block)
       Nokogiri::HTML.fragment(@rendered_component)
     end
 
-    def controller
-      @controller ||= Base.test_controller.constantize.new.tap { |c| c.request = request }.extend(Rails.application.routes.url_helpers)
-    end
-
     def request
       @request ||= ActionDispatch::TestRequest.create
     end
 
-    def with_variant(variant)
-      old_variants = controller.view_context.lookup_context.variants
+    def with_variant(new_variant)
+      old_variant = variant
 
-      controller.view_context.lookup_context.variants = variant
+      @variant = new_variant
       yield
-      controller.view_context.lookup_context.variants = old_variants
+      @variant = old_variant
+    end
+
+    private
+
+    def build_controller(component)
+      component.test_controller.constantize.new.tap { |c| c.request = request }.extend(Rails.application.routes.url_helpers)
```

Comment:
I'm imagine that the `constantize` is slightly expensive and we aren't memoizing it per component. I wonder if we can arrange that?

Reply:
Definitely can't hurt, done!